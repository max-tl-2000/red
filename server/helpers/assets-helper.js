/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import path from 'path';
import fs from 'fs';
import { DALTypes } from '../../common/enums/DALTypes';
import { formatEntityAssetUrl } from './tenantContextConfigs';
import loggerModule from '../../common/helpers/logger';
const logger = loggerModule.child({ subType: 'assetsHelper' });

// this regular expression is not understood by firefox
// by moving it here (server side), where it is used it fixes the issue
const ASSET_TAGS_BETWEEN_HYPHENS = /(?<=-)([a-z]{2})(?=-)/gi;
const ASSET_RANK_LABEL = /(\d{1,3})\s*-{0,1}\s*(.*)/i;
const ASSET_URL_SCAPE_CHARACTERS = /[\s#$%&@`\/:;<=>?[\\\]^{|}~"'+,]/gi; // eslint-disable-line no-useless-escape

const extractCharacters = (searchExpression, indexToExtract) => text => {
  let textExtracted;
  const result = new Set();

  do {
    textExtracted = searchExpression.exec(text);
    textExtracted && result.add(textExtracted[indexToExtract]);
  } while (textExtracted);
  return Array.from(result);
};

const isTagOfType = type => tag => !!tag && tag.toUpperCase() === type.toUpperCase();

export const parseMetadataInFilePath = filePath => {
  const fileName = path.parse(filePath).name;
  const extractTags = extractCharacters(ASSET_TAGS_BETWEEN_HYPHENS, 1);

  const tags = extractTags(fileName);
  const tagsInName = tags.length ? `${tags.join('-')}-` : '';
  const [, rankExtracted, labelExtracted = ''] = fileName.match(ASSET_RANK_LABEL) || [];

  const highValue = tags.some(isTagOfType('HV'));
  const floorPlan = tags.some(isTagOfType('FP'));
  const rank = rankExtracted ? parseInt(rankExtracted, 10) : 0; // no ranking information
  const label = labelExtracted.substring(labelExtracted.indexOf(tagsInName) + tagsInName.length);

  return { highValue, floorPlan, rank, label };
};

export const parseMetadataInRxpFilePath = (ctx, filePath) => {
  const fileName = path.parse(filePath).name;
  const fileNameParts = fileName.split('-'); // ex: 64x64-square-dark.png

  if (fileNameParts.length < 3) {
    logger.error({ ctx, filePath }, 'Rxp asset name does not contain the required info');
    throw new Error('INCORRECT_RXP_ASSET_NAME');
  }

  const themeNameParts = fileNameParts[fileNameParts.length - 1].split('.');

  return {
    dimensions: fileNameParts[0],
    shape: fileNameParts[1],
    theme: themeNameParts[0], // without file extension
  };
};

export const validateMarketingAsset = (filePath, { fileSize = 2.5, fileNameLength = 255 } = {}) => {
  const fileName = path.parse(filePath).base;
  const errors = [];

  const extractInvalidCharacters = extractCharacters(ASSET_URL_SCAPE_CHARACTERS, 0);
  const invalidCharacters = extractInvalidCharacters(fileName);
  invalidCharacters.length && errors.push({ token: 'INVALID_CHARACTERS', message: invalidCharacters.join('') });

  fileName.length > fileNameLength && errors.push({ token: 'INVALID_FILENAME_SIZE', message: `Limited to ${fileNameLength} charcaters` });

  const { size } = fs.statSync(filePath);
  const maxSizeInBytes = fileSize * 1000 * 1000;
  size > maxSizeInBytes && errors.push({ token: 'INVALID_FILE_SIZE', message: `Limited to ${fileSize} megabytes` });

  return { success: !errors.length, errors };
};

export const formatEmployeeAssetUrl = async (ctx, entityId, options) => await formatEntityAssetUrl(ctx, entityId, DALTypes.AssetType.EMPLOYEE, options);

export const formatInventoryAssetUrl = async (ctx, entityId, options) => await formatEntityAssetUrl(ctx, entityId, DALTypes.AssetType.INVENTORY, options);

export const formatPropertyAssetUrl = async (ctx, entityId, options) => await formatEntityAssetUrl(ctx, entityId, DALTypes.AssetType.PROPERTY, options);
