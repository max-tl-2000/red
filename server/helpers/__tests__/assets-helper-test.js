/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import path from 'path';
import { parseMetadataInFilePath, parseMetadataInRxpFilePath, validateMarketingAsset } from '../assets-helper';

describe('assetsHelper', () => {
  describe('when parseMetadataInFilePath is called', () => {
    it('should extract metadata (highValue, floorPlan, rank, label)', () => {
      const resultMetadata = [];
      ['01-FP-HV-some name.ext', '01-Office.jpg', '02-FP-1x1 TH 691-799 SF.png', '01 - Living dining.jpg', '02-HV-Patio.jpg', 'HV-Patio.jpg'].forEach(
        fileName => {
          const parsed = parseMetadataInFilePath(fileName);
          resultMetadata.push({ fileName, ...parsed });
        },
      );
      expect(resultMetadata).toMatchSnapshot();
    });
  });

  describe('when parseMetadataInRxpFilePath is called', () => {
    describe('and filename is correct', () => {
      it('should extract metadata (dimensions, shape, theme)', () => {
        const resultMetadata = [];
        ['64x64-square-dark.png', '64x64-square-dark.jpg', '32x32-square-dark.png', '32x32-circle-dark.png', '32x32-circle-light.png'].forEach(fileName => {
          const parsed = parseMetadataInRxpFilePath({}, fileName);
          resultMetadata.push(parsed);
        });
        expect(resultMetadata).toMatchSnapshot();
      });
    });

    describe('and filename is incorrect', () => {
      it('should throw error: INCORRECT_RXP_ASSET_NAME', () => {
        const result = [];
        ['square-dark.png', '64x64.square.dark.jpg', '32x32.png'].forEach(fileName => {
          try {
            parseMetadataInRxpFilePath({}, fileName);
          } catch (ex) {
            result.push(ex.message);
          }
        });
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('when validateMarketingAsset is called', () => {
    [
      {
        fileName: 'Dylan_BlackLogo.png',
        result: {
          success: true,
          errors: [],
        },
      },
      {
        fileName: 'invalid-file-size.png.png',
        result: {
          success: false,
          errors: [
            {
              token: 'INVALID_FILE_SIZE',
              message: 'Limited to 2.5 megabytes',
            },
          ],
        },
      },
      {
        fileName: 'invalid file %$@| name.png',
        result: {
          success: false,
          errors: [
            {
              token: 'INVALID_CHARACTERS',
              message: ' %$@|',
            },
          ],
        },
      },
      {
        fileName: 'invalid-file-name-size-and-invalid-characters #d.png',
        result: {
          success: false,
          errors: [
            {
              token: 'INVALID_CHARACTERS',
              message: ' #',
            },
            {
              token: 'INVALID_FILENAME_SIZE',
              message: 'Limited to 50 charcaters',
            },
          ],
        },
      },
    ].forEach(data => {
      it(`should validate ${data.fileName} returning valid ${data.result.success}`, () => {
        const filePath = path.resolve(path.dirname(__dirname), '__tests__', 'assets', data.fileName);
        const result = validateMarketingAsset(filePath, { fileNameLength: 50 });
        expect(result.success).toEqual(data.result.success);
        expect(result).toEqual(data.result);
      });
    });
  });
});
