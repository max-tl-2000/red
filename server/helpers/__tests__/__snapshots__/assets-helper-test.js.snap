// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assetsHelper when parseMetadataInFilePath is called should extract metadata (highValue, floorPlan, rank, label) 1`] = `
Array [
  Object {
    "fileName": "01-FP-HV-some name.ext",
    "floorPlan": true,
    "highValue": true,
    "label": "some name",
    "rank": 1,
  },
  Object {
    "fileName": "01-Office.jpg",
    "floorPlan": false,
    "highValue": false,
    "label": "Office",
    "rank": 1,
  },
  Object {
    "fileName": "02-FP-1x1 TH 691-799 SF.png",
    "floorPlan": true,
    "highValue": false,
    "label": "1x1 TH 691-799 SF",
    "rank": 2,
  },
  Object {
    "fileName": "01 - Living dining.jpg",
    "floorPlan": false,
    "highValue": false,
    "label": "Living dining",
    "rank": 1,
  },
  Object {
    "fileName": "02-HV-Patio.jpg",
    "floorPlan": false,
    "highValue": true,
    "label": "Patio",
    "rank": 2,
  },
  Object {
    "fileName": "HV-Patio.jpg",
    "floorPlan": false,
    "highValue": false,
    "label": "",
    "rank": 0,
  },
]
`;

exports[`assetsHelper when parseMetadataInRxpFilePath is called and filename is correct should extract metadata (dimensions, shape, theme) 1`] = `
Array [
  Object {
    "dimensions": "64x64",
    "shape": "square",
    "theme": "dark",
  },
  Object {
    "dimensions": "64x64",
    "shape": "square",
    "theme": "dark",
  },
  Object {
    "dimensions": "32x32",
    "shape": "square",
    "theme": "dark",
  },
  Object {
    "dimensions": "32x32",
    "shape": "circle",
    "theme": "dark",
  },
  Object {
    "dimensions": "32x32",
    "shape": "circle",
    "theme": "light",
  },
]
`;

exports[`assetsHelper when parseMetadataInRxpFilePath is called and filename is incorrect should throw error: INCORRECT_RXP_ASSET_NAME 1`] = `
Array [
  "INCORRECT_RXP_ASSET_NAME",
  "INCORRECT_RXP_ASSET_NAME",
  "INCORRECT_RXP_ASSET_NAME",
]
`;
